File: home/.claude/skills/news-briefing/extract_article.py
#!/usr/bin/env python3
"""
Extract article body text from a news article's HTML.
Reads HTML from stdin.

Usage: curl -s "ARTICLE_URL" | python3 extract_article.py
Output: article paragraphs as plain text, capped at ~3000 characters.
"""

import sys
import re

html = sys.stdin.buffer.read().decode("utf-8", errors="replace")

# Prefer content inside an <article> tag if present
article = re.search(r"<article\b[^>]*>(.*?)</article>", html, re.DOTALL)
content = article.group(1) if article else html

total = 0
for p in re.findall(r"<p\b[^>]*>(.*?)</p>", content, re.DOTALL):
    text = re.sub(r"<[^>]+>", " ", p).strip()
    text = re.sub(r"\s+", " ", text)
    if len(text) > 40:
        print(text)
        total += len(text)
        if total > 3000:
            break

File: home/.claude/skills/news-briefing/extract_article.ts
#!/usr/bin/env node

/** Extract article paragraphs from a news page's HTML. */
function extractArticleParagraphs(
  /** The raw HTML to parse. */
  html: string,
): string[] {
  const articleMatch = html.match(/<article\b[^>]*>(.*?)<\/article>/is)
  const content = articleMatch?.[1] ?? html
  let total = 0
  const paragraphs: string[] = []

  for (const match of content.matchAll(/<p\b[^>]*>(.*?)<\/p>/gis)) {
    const text = match[1]
      .replace(/<[^>]+>/g, " ")
      .trim()
      .replace(/\s+/g, " ")

    if (text.length <= 40) {
      continue
    }

    paragraphs.push(text)
    total += text.length

    if (total > 3000) {
      break
    }
  }

  return paragraphs
}

const { readFileSync } = require("node:fs")

const html = readFileSync(0, "utf8")
const paragraphs = extractArticleParagraphs(html)
process.stdout.write(paragraphs.join("\n"))

if (paragraphs.length > 0) {
  process.stdout.write("\n")
}
